### PR TITLE
BUG: Fix #26210: MultiIndex Sorting with NaNs

### LINKED ISSUE #26210: BUG: Sorting MultiIndex with NaNs
#### Code Sample, a copy-pastable example if possible

```python
>>> ix = pd.MultiIndex(levels=[['B', 'A'], ['x']], codes=[[-1, 0, 1], [0, 0, 0]],
...             names=['symbol', 'date'])
>>> ix._sort_levels_monotonic()
MultiIndex(levels=[['A', 'B'], ['x']],
           codes=[[0, 1, 0], [0, 0, 0]],
           names=['symbol', 'date'])

# MultiIndex silently accepts Nan in codes when inside a NumPy array and cast to 0:

>>> pd.MultiIndex(levels=[['B', 'A'], ['x']], codes=[np.array([np.nan, 0, 1]), [0, 0, 0]],
...             names=['symbol', 'date'])
MultiIndex(levels=[['B', 'A'], ['x']],
           codes=[[0, 0, 1], [0, 0, 0]],
           names=['symbol', 'date'])
```
#### Problem description

When sorting a `MultiIndex`, NaNs in any level which is not already sorted are replaced with the first non-null value in the sorted orders. Effectively, the `-1` values in the `codes` get replaced with `0`. I've tracked this to two separate parts. The first part is in this line in `MultiIndex._sort_levels_monotonic`:

                    level_codes = algos.take_1d(ri, level_codes)

The output has `np.nan` where the input had `-1`. This could be fixed by:

                    level_codes = algos.take_1d(ri, level_codes, fill_value=-1)

The second part, which also seems unexpected to me, is that `MultiIndex` silently accepts NaN in codes when inside a NumPy array and casts to 0 (see second part of example above).
In particular, this is inconsistent with the behaviour when we pass NaNs inside a `list` which throws an error:

```python

>>> pd.MultiIndex(levels=[['B', 'A'], ['x']], codes=[[np.nan, 0, 1], [0, 0, 0]])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/dtc/conda_envs/condron/envs/pd24/lib/python3.6/site-packages/pandas/util/_decorators.py", line 188, in wrapper
    return func(*args, **kwargs)
  File "/opt/dtc/conda_envs/condron/envs/pd24/lib/python3.6/site-packages/pandas/core/indexes/multi.py", line 235, in __new__
    result._set_codes(codes, copy=copy, validate=False)
  File "/opt/dtc/conda_envs/condron/envs/pd24/lib/python3.6/site-packages/pandas/core/indexes/multi.py", line 683, in _set_codes
    for lev, level_codes in zip(self.levels, codes))
  File "/opt/dtc/conda_envs/condron/envs/pd24/lib/python3.6/site-packages/pandas/core/indexes/multi.py", line 683, in <genexpr>
    for lev, level_codes in zip(self.levels, codes))
  File "/opt/dtc/conda_envs/condron/envs/pd24/lib/python3.6/site-packages/pandas/core/indexes/frozen.py", line 192, in _ensure_frozen
    array_like = coerce_indexer_dtype(array_like, categories)
  File "/opt/dtc/conda_envs/condron/envs/pd24/lib/python3.6/site-packages/pandas/core/dtypes/cast.py", line 556, in coerce_indexer_dtype
    return ensure_int8(indexer)
  File "pandas/_libs/algos_common_helper.pxi", line 333, in pandas._libs.algos.ensure_int8
ValueError: cannot convert float NaN to integer

```

#### Expected Output

I think the expected output of `_sort_levels_monotonic` in this case would preserve the `-1` codes.  Patching `_sort_levels_monotonic` with this fix I get:

````python

>>> s = pd.Series([1, 2, 3], ix)
>>> s.sort_index(na_position='first')
symbol  date
NaN     x       1
A       x       3
B       x       2

>>> s.sort_index(na_position='last')
symbol  date
A       x       3
B       x       2
NaN     x       1

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.162-94.72-default
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.24.2
pytest: None
pip: 9.0.1
setuptools: 36.4.0
Cython: None
numpy: 1.13.1
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by eoincondron**
- [x] closes #xxxx
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -1707,7 +1707,7 @@ def remove_unused_levels(self):
                 # ... and reassigned value -1:
                 code_mapping[uniques] = np.arange(len(uniques)) - has_na
 
-                level_codes = code_mapping[level_codes]
+                level_codes = code_mapping[level_codes].astype(int)
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: why would u need to do this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (eoincondron) [reply pandas/core/indexes/multi.py]: This is necessary given the update to `coerce_indexer_dtype` (it breaks some tests of this method). I wasn't sure about this change, I mentioned it in the initial report. The issue could be handled more locally in the `MultiIndex._set_codes`. But it seems to me you don't really want to pass `floats` to `categorical` indexers (which is happening here). Looking at it now, we should probably do `code_mapping[uniques] = np.arange(len(uniques), dtype=int)` above.

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -564,6 +564,9 @@ def invalidate_string_dtypes(dtype_set):
 
 def coerce_indexer_dtype(indexer, categories):
     """ coerce the indexer input array to the smallest dtype possible """
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: best to add a doc-string here; why are you adding these changes?

File: pandas/tests/indexes/multi/test_constructor.py [hunk 3]
```diff
@@ -107,6 +107,13 @@ def test_copy_in_constructor():
     assert mi.levels[0][0] == val
 
 
+def test_constructor_bad_codes():
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_constructor.py]: bad -> invalid

add the issue number as a comment

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/multi.py]: I don't want to interject a random astype in a particular method; this copies, and it is very fragile

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -564,6 +564,9 @@ def invalidate_string_dtypes(dtype_set):
...
+    indexer = np.array(indexer)
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: use np.asarray; though see my other comment; this is an internal method; these are guaranteed to *already* be an integer index, if not then there is a problem in the caller.

File: pandas/tests/indexing/multiindex/test_sorted.py [hunk 5]
```diff
@@ -90,3 +91,23 @@ def test_series_getitem_not_sorted(self):
         expected.index = expected.index.droplevel(0)
         tm.assert_series_equal(result, expected)
         tm.assert_series_equal(result2, expected)
+
```
Comment (jreback) [inline pandas/tests/indexing/multiindex/test_sorted.py]: these should be here: pandas/tests/indexes/multi/test_sorting.py:

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
does this have an issue? merge master as well

File: pandas/tests/indexing/multiindex/test_sorted.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import pytest
```
Comment (jreback) [inline pandas/tests/indexing/multiindex/test_sorted.py]: is this needed?

Comment (eoincondron) [reply pandas/tests/indexing/multiindex/test_sorted.py]: No, I've remove that now.

File: pandas/tests/indexing/multiindex/test_sorted.py (hunk 5 above)
Comment (eoincondron) [reply pandas/tests/indexing/multiindex/test_sorted.py]: I've moved them across.

File: pandas/tests/indexes/multi/test_constructor.py (hunk 3 above)
Comment (eoincondron) [reply pandas/tests/indexes/multi/test_constructor.py]: Done.

File: pandas/core/dtypes/cast.py (hunk 2 above)
Comment (eoincondron) [reply pandas/core/dtypes/cast.py]: Done.

File: pandas/core/dtypes/cast.py (hunk 4 above)
Comment (eoincondron) [reply pandas/core/dtypes/cast.py]: Thanks for the feedback. I've used `asarray` as you suggest. When you say they are guaranteed to be an integer index already, do you mean that this should be guaranteed by ensuring this contract is upheld every us case within the codebase? I can take it out if you prefer. The case where this contract was broken in `remove_unused_levels` has been resolved.

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (eoincondron) [reply pandas/core/indexes/multi.py]: I've replace this with `code_mapping[uniques] = np.arange(len(uniques), dtype=int) - has_na` which I should have done initially. This seems to natural to me as ensure `code_mapping` is of the correct type to begin with.

**Comment by jreback** [comment]
can you merge master and see if can get passing

**Comment by jreback** [comment]
closing as stale, if you'd like to continue, pls ping.

**Pull request closed without merging**